File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} list1
     * @param {ListNode} list2
     * @return {ListNode}
     */
    mergeTwoLists(list1, list2) {

        // **** Recurring method **********
        // if(!list1) {
        //     return list2
        // }

        // if(!list2) return list1

        // if(list1.val < list2.val) {
        //     list1.next = this.mergeTwoLists(list1.next , list2)
        //     return list1
        // } else {
        //      list2.next = this.mergeTwoLists(list1 , list2.next)
        //     return list2
        // }

        let newNode = {val:0 , next:null}
        let dummy = newNode

        while(list1 && list2) {
            if(list1.val < list2.val) {
                dummy.next = list1
                list1 = list1.next
            } else {
                dummy.next = list2
                list2 = list2.next
            }

            dummy = dummy.next
        }

        if(list1) {
            dummy.next = list1
        } else {
            dummy.next = list2
        }

        return newNode.next

    }
}

File: Data Structures & Algorithms/add-two-numbers/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        
        int carry = 0;
        ListNode dummyNode = new ListNode();
        ListNode currD = dummyNode;

        while(l1 != null || l2 != null || carry != 0) {

            int val1 = (l1 != null) ? l1.val : 0;
            int val2 = l2 != null ? l2.val  : 0;

            int sum = val1 + val2 + carry;
            carry = sum / 10; // java automatically take floor value 

            int val  = sum % 10;
            
            currD.next = new ListNode(val);

            currD = currD.next;

            l1 = (l1 != null) ? l1.next : null;
            l2 = (l2 != null) ? l2.next : null;
        



        }

        return dummyNode.next;
    }
}
